lib/logic/player/favorites.dart

```diff
@@ -1,27 +1,133 @@
+          if (song.isFavoriteInMediaStore!) {
+            favoriteSet.add(song.sourceId);
+          }
+        }
+      } else {
+        final serializer = repository.serializersMap.getValue(contentType);
+        await serializer!.init();
+        final savedIds = await serializer.read();
+        favoriteSet.addAll(savedIds);
+      }
+      _favoriteSetsMap.setValue(favoriteSet, key: contentType);
+    }
+  }
+
+  @override
+  void dispose() {
+    _favoriteSetsMap.clear();
+    super.dispose();
+  }
+
+  ValueListenable<bool> get onFavoritesMode => _favoritesModeNotifier;
+  final ValueNotifier<bool> _favoritesModeNotifier = ValueNotifier(false);
+
+  /// Whether tabs route currently should filter and only show favorite content.
+  bool get favoritesMode => _favoritesModeNotifier.value;
+
+  void switchFavoritesMode() {
+    _favoritesModeNotifier.value = !_favoritesModeNotifier.value;
+  }
+
+  /// Whether the given [content] is favorite.
+  bool isFavorite<T extends Content>(T content) {
+    final favoriteSet = _favoriteSetsMap.getValue<T>(content.runtimeType);
+    if (content is Song)
+      return favoriteSet!.contains(content.sourceId);
+    return favoriteSet!.contains(content.id);
+  }
+
+  /// Sets whether a given tuple of content is favorite.
+  Future<void> setFavorite({
+    required ContentTuple contentTuple,
+    required bool value
+  }) async {
+    for (final contentType in Content.enumerate()) {
+      final contentList = contentTuple.get(contentType);
+      final newFavoriteSet = _favoriteSetsMap.getValue(contentType)!.toSet();
+      Iterable<int> ids;
+      if (contentType == Song) {
+        ids = (contentList as List<Song>).map((el) => el.sourceId);
+      } else {
+        ids = contentList.map((el) => el.id);
+      }
+
+      if (ids.isEmpty) {
+        continue;
+      }
+
+      if (value) {
+        newFavoriteSet.addAll(ids);
+      } else {
+        newFavoriteSet.removeAll(ids);
+      }
+
+      try {
+        if (_useAndroidFavorites(contentType)) {
+          final songs = (contentList as List<Song>).toSet();
+          await ContentControl.instance.setSongsFavorite(songs, value);
+        } else {
+          final serializer = repository.serializersMap.getValue(contentType);
+          await serializer!.save(newFavoriteSet.toList());
+        }
+        _favoriteSetsMap.setValue(newFavoriteSet, key: contentType);
+        ContentControl.instance.emitContentChange();
+        QueueControl.instance.emitQueueChange();
+      } catch (ex, stack) {
+        FirebaseCrashlytics.instance.recordError(
+          ex,
+          stack,
+          reason: 'in setFavorite',
+        );
+        ShowFunctions.instance.showToast(
+          msg: staticl10n.oopsErrorOccurred,
+        );
+        debugPrint('setFavorite error: $ex');
+      }
+    }
+  }
+
+  /// Switches current song favorite status.
+  Future<void> switchFavoriteCurrentSong() {
```

Comment:
toggle is better, but I feel like it should make it clear that it works with currentSong

Reply:
But if you move this function to the `Song` class and replace `currentSong` with `this` it would work for any song.
For the current song you can then write
```dart
PlaybackControl.instance.currentSong.toggleFavourite();
```